protocols/autonatv2/src/server/behaviour.rs

```diff
@@ -0,0 +1,235 @@
+    fn default() -> Self {
+        Self::new(OsRng)
+    }
+}
+
+impl<R> Behaviour<R>
+where
+    R: RngCore + Send + Clone + 'static,
+{
+    pub fn new(rng: R) -> Self {
+        Self {
+            handlers: HashMap::new(),
+            dialing_dial_back: HashMap::new(),
+            pending_events: VecDeque::new(),
+            rng,
+        }
+    }
+
+    fn poll_pending_events(
+        &mut self,
+        _cx: &mut Context<'_>,
+    ) -> Poll<
+        ToSwarm<
+            <Self as NetworkBehaviour>::ToSwarm,
+            <<Self as NetworkBehaviour>::ConnectionHandler as ConnectionHandler>::FromBehaviour,
+        >,
+    > {
+        if let Some(event) = self.pending_events.pop_front() {
+            return Poll::Ready(event);
+        }
+        Poll::Pending
+    }
+}
+
+impl<R> NetworkBehaviour for Behaviour<R>
+where
+    R: RngCore + Send + Clone + 'static,
+{
+    type ConnectionHandler = Handler<R>;
+
+    type ToSwarm = ();
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        _connection_id: ConnectionId,
+        _peer: PeerId,
+        _local_addr: &Multiaddr,
+        remote_addr: &Multiaddr,
+    ) -> Result<<Self as NetworkBehaviour>::ConnectionHandler, ConnectionDenied> {
+        Ok(Either::Right(dial_request::Handler::new(
+            remote_addr.clone(),
+            self.rng.clone(),
+        )))
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        peer: PeerId,
+        addr: &Multiaddr,
+        _role_override: Endpoint,
+        port_use: PortUse,
+    ) -> Result<<Self as NetworkBehaviour>::ConnectionHandler, ConnectionDenied> {
+        if port_use == PortUse::New {
+            self.handlers.insert(
+                (
+                    addr.iter()
+                        .filter(|e| !matches!(e, Protocol::P2p(_)))
+                        .collect(),
+                    peer,
+                ),
+                connection_id,
+            );
+        }
+        Ok(Either::Left(dial_back::Handler::new()))
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm) {
+        match event {
+            FromSwarm::DialFailure(DialFailure {
+                error: DialError::WrongPeerId { .. },
+                peer_id: Some(peer_id),
+                ..
+            }) => {
+                if let Some(key) = self
+                    .dialing_dial_back
+                    .keys()
+                    .find(|(_, p)| *p == peer_id)
+                    .cloned()
+                {
+                    let cmds = self.dialing_dial_back.remove(&key).unwrap();
+                    for cmd in cmds {
+                        let _ = cmd.back_channel.send(DialBack::Dial);
+                    }
+                }
+            }
+            FromSwarm::DialFailure(DialFailure {
+                error: DialError::Transport(pairs),
+                ..
+            }) => {
+                for (addr, _) in pairs.iter() {
```

Comment:
It might be helpful to either log the error here _or_ (perhaps even better), send the error through the channel as well to tell the handler, why the dial failed so we can handle all that in one place.

Reply:
I think that is now unnecessary as well.